In [1]:
import numpy as np
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score

train_features = np.array([[0.346,0.780],[0.303,0.439],[0.358,0.729],[0.602,0.863],[0.790,0.753],[0.611,0.965]])
train_labels = np.array([0,0,0,1,1,1])
test_features = np.array([[0.959,0.382],[0.750,0.306],[0.395,0.760],[0.823,0.764],[0.761,0.874],[0.844,0.435]])
test_labels = np.array([0,0,0,1,1,1])
train_features = np.column_stack([np.ones(len(train_features)),train_features])
test_features = np.column_stack([np.ones(len(test_features)),test_features])

In [2]:
"""PART A"""

class LogisticRegressionClassifier():
    
    def __init__(self):
        self.theta = np.array([-1,1.5,0.5])
        self.eta = 0.1
        self.tolerance = 0.01
        self.epoch = 1
        
    def sigmoid(self,record):
        return 1/(1+np.exp(-self.theta@record))
        
    def learn(self,train_features,train_labels):
        gradient = np.array([10000,10000,10000])
        while np.linalg.norm(gradient) > self.tolerance:
            gradient = 0
            for i in range(len(train_features)):
                gradient += train_features[i,:]*(train_labels[i]-self.sigmoid(train_features[i,:]))
            self.theta += self.eta*gradient/6
            if self.epoch == 1:
                print("weights after one iteration =", self.theta)
                self.epoch += 1
            
    def classify(self,record):
        if self.sigmoid(record) > 0.5:
            return 1
        else:
            return 0

The above cell implements the logistic regression classifier required in part a of the question.

Part b:

The probabilty of a record being 1 at the beginning is given by
\begin{equation}
    P(\hat{y} = 1|x_1,x_2) = \dfrac{1}{1+e^{-(-1+1.5x_1+0.5x_2)}}
\end{equation}
The cross entropy error function is given by 
\begin{equation}
    E(w) = \sum_ly^llnP(y^l=1|x^l,w)+(1-y^l)lnP(y^l=0|x^l,w)
\end{equation}
where w = [-1, 1.5, 0.5]

In [3]:
clf = LogisticRegressionClassifier()
clf.learn(train_features,train_labels)

preds = np.zeros(6)
for i in range(6):
    preds[i] = clf.classify(test_features[i,:])

print("Accuracy =", accuracy_score(test_labels, preds))
print("Recall =", recall_score(test_labels, preds))
print("Precision =", precision_score(test_labels, preds))

weights after one iteration = [-1.00316626  1.50535086  0.50196867]
Accuracy = 0.6666666666666666
Recall = 1.0
Precision = 0.6


The updated logistic model after one iteration is
\begin{equation}
    P(\hat{y} = 1|x_1,x_2) = \dfrac{1}{1+e^{-(-1.0032+1.5053x_1+0.502x_2)}}
\end{equation}

After the model converges, Accuracy = 0.67, Recall = 1.0, Precision = 0.6